clarity/contracts/arkadiko-swap-v1-1.clar

```diff
@@ -211,8 +232,88 @@
-  (ok (map get-pair-contracts (var-get pairs-list)))
+(define-public (migrate-create-pair (token-x-trait <ft-trait>) (token-y-trait <ft-trait>) (swap-token-trait <swap-token>) (pair-name (string-ascii 32)) (shares-total uint))
+
+  (let (
+    (pair-id (+ (var-get pair-count) u1))
+    (token-x (contract-of token-x-trait))
+    (token-y (contract-of token-y-trait))
+    (token-swap (contract-of swap-token-trait))
+
+    (pair-data {
+      enabled: false,
+      shares-total: shares-total,
+      balance-x: u0,
+      balance-y: u0,
+      fee-balance-x: u0,
+      fee-balance-y: u0,
+      fee-to-address: (some (contract-call? .arkadiko-dao get-payout-address)),
+      swap-token: token-swap,
+      name: pair-name,
+    })
+  )
+    (asserts! (is-eq contract-caller (contract-call? .arkadiko-dao get-dao-owner)) (err ERR-NOT-AUTHORIZED))
+    (asserts!
+      (and
+        (is-none (map-get? pairs-data-map { token-x: token-x, token-y: token-y }))
+        (is-none (map-get? pairs-data-map { token-x: token-y, token-y: token-x }))
+      )
+      pair-already-exists-err
+    )
+
+    ;; Register swap token
+    (try! (register-swap-token token-swap))
+
+    ;; Update maps
+    (map-set pairs-data-map { token-x: token-x, token-y: token-y } pair-data)
+    (map-set pairs-map { pair-id: pair-id } { token-x: token-x, token-y: token-y })
+
+    ;; Increase pair count
+    (var-set pair-count pair-id)
+
+    (ok true)
+  )
+
+)
+
+(define-public (migrate-add-liquidity (token-x-trait <ft-trait>) (token-y-trait <ft-trait>) (x uint) (y uint))
+
+  (let (
+    (contract-address (as-contract tx-sender))
+    (token-x (contract-of token-x-trait))
+    (token-y (contract-of token-y-trait))
+    (pair-data (unwrap-panic (map-get? pairs-data-map { token-x: token-x, token-y: token-y })))
+    (new-balance-x (+ (get balance-x pair-data) x))
+    (new-balance-y (+ (get balance-y pair-data) y))
+  )
+    (asserts! (is-eq contract-caller (contract-call? .arkadiko-dao get-dao-owner)) (err ERR-NOT-AUTHORIZED))
+
+    ;; Transfer tokens from tx-sender to this contract
+    (if (is-eq token-x .wrapped-stx-token)
+      (begin
+        (try! (contract-call? .arkadiko-dao mint-token .wrapped-stx-token x tx-sender))
+        (try! (stx-transfer? x tx-sender contract-address))
+      )
+      false
+    )
+    (if (is-eq token-y .wrapped-stx-token)
+      (begin
+        (try! (contract-call? .arkadiko-dao mint-token .wrapped-stx-token y tx-sender))
+        (try! (stx-transfer? y tx-sender contract-address))
+      )
+      false
+    )
+    (asserts! (is-ok (contract-call? token-x-trait transfer x tx-sender contract-address none)) transfer-x-failed-err)
+    (asserts! (is-ok (contract-call? token-y-trait transfer y tx-sender contract-address none)) transfer-y-failed-err)
+
+    (map-set pairs-data-map
+      { token-x: token-x, token-y: token-y }
+      (merge pair-data { balance-x: new-balance-x, balance-y: new-balance-y })
+    )
+
+    (ok true)
+  )
```

Comment:
Is there a way to design the migration differently with a free `begin` statement, executed once when the contract is being deployed, that would fetch the data from the swap v1?
I don't feel particularly good with the idea of adding more public function (I can see that these are guarded tho, but still).

Reply:
@nieldeckx thoughts?

I think fetching from the v1 contract would be hard tho.. all the funds are also on a contract we own ATM.

Reply:
Alternatively we could also allow doing this only up until a certain block height.